src/node/wsRouter.ts

```diff
@@ -0,0 +1,57 @@
+import * as express from "express"
+import * as expressCore from "express-serve-static-core"
+import * as http from "http"
+import * as net from "net"
+
+export const handleUpgrade = (app: express.Express, server: http.Server): void => {
+  server.on("upgrade", (req, socket, head) => {
+    socket.on("error", () => socket.destroy())
+
+    req.ws = socket
+    req.head = head
+    req._ws_handled = false
+
+    // Send the request off to be handled by Express.
+    ;(app as any).handle(req, new http.ServerResponse(req), () => {
+      if (!req._ws_handled) {
+        socket.destroy(new Error("Not found"))
```

Comment:
`end`.

Reply:
Shouldn't we write a 404 here actually?

Reply:
I'm a bit nervous with `end` because that's what I used for the VS Code socket and it resulted in the connections lingering causing the heartbeat file to be inaccurate: https://github.com/cdr/code-server/commit/9ff0e455c366449b2ba02674d878303fb967ef36

Maybe this was a symptom of some other problem though? No idea. I think maybe it's possible to keep the socket open even after `end` in some cases but I don't know exactly what.

Can we write a 404 to a socket? Node gives them to us already upgraded. My knowledge around how exactly this all works is pretty fuzzy.

Reply:
Yea so we're always writing a 101 above right. We shouldn't be doing that. Only when we find a handler should the 101 be written, otherwise it should be a normal 404.

If that's not clear, see https://tools.ietf.org/html/rfc6455#section-1.2 for details on how the handshake process works

Reply:
I think I just don't know what I'm talking about lol. We handle the upgrade, not Node. I think I have my terminology mixed up. So we can just `socket.write` the 404 if I understand correctly.

Reply:
I think this is sorted now. We only 101 on the VS Code endpoint, otherwise if it's not handled it's a 404.